### PR TITLE
legend markers and colors for scatter plot

### PR DESCRIPTION
**Pull request by moontrip**
Depends on the PR at web-components, [#253](https://github.com/VEuPathDB/web-components/pull/253)

This PR addresses #678  and #679, which are related to custom legend for scatter plot viz. 

A screenshot is attached:

scatterplot-legend-marker-example1

### CODE REVIEW

File: src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx [hunk 1]
```diff
@@ -89,6 +89,7 @@ import PlotLegend, {
 } from '@veupathdb/components/lib/components/plotControls/PlotLegend';
 import { isFaceted } from '@veupathdb/components/lib/types/guards';
 import FacetedPlot from '@veupathdb/components/lib/plots/FacetedPlot';
+import { config } from 'process';
```
Comment (jtlong3rd) [inline src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx:92]: Oops! I'm guessing you don't need this import?

File: src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx [hunk 2]
```diff
@@ -388,18 +389,68 @@ function ScatterplotViz(props: VisualizationProps) {
...
+      ? // the name 'dataItem' is used inside the map() to distinguish from the global 'data' variable
+        legendData.map((dataItem: XYPlotDataSeries, index: number) => {
           return {
-            label: data.name != null ? data.name : '',
-            // TO-do: need a way to appropriately make marker info
-            // scatter plot has defined mode - still need to distinguish CI, No data, etc.
-            marker: data.mode != null ? data.mode : '',
-            // set marker colors
-            markerColor: 'markerColor',
+            label: dataItem.name != null ? dataItem.name : '',
```
Comment (jtlong3rd) [inline src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx:416]:
```suggestion
            label: dataItem.name ?? '',
```

File: src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx [hunk 3]
```diff
@@ -388,18 +389,68 @@ function ScatterplotViz(props: VisualizationProps) {
...
+            marker:
+              dataItem.mode != null
+                ? dataItem.name === 'No data'
+                  ? 'x'
+                  : dataItem.mode === 'markers'
+                  ? 'circle'
+                  : dataItem.mode === 'lines'
+                  ? 'line'
+                  : ''
+                : dataItem.fill != null && dataItem.fill === 'tozeroy'
```
Comment (jtlong3rd) [inline src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx:427]: Maybe something like

```suggestion
                : dataItem?.fill === 'tozeroy'
```

**Review by jtlong3rd — APPROVED** [review]
Save for some small inline comments, this should be good to merge!

File: src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx (hunk 1 above)
Comment (moontrip) [reply src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx:92]: @jtlong3rd Thank you for your review and suggestions! 👍 IDK why it was entered but perhaps due to a sort of VS Code's function which is often not that smart 😄

File: src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx (hunk 2 above)
Comment (moontrip) [reply src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx:416]: Yep will do

File: src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx (hunk 3 above)
Comment (moontrip) [reply src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx:427]: Yep!

**Review by bobular — CHANGES_REQUESTED** [review]
Looks good.  Nice colormap solution to the tricky problem of all the series being interleaved.

Might be nice to put `', Smoothed mean'`, `', 95% Confidence interval'` etc into "global" constants for this file, so that future edits will only need to be done in one place.

There is potentially a problem with some real-life variables having actual "No data" values from the database/back end.  I don't know one off the top of my head, but will ask Sheena/Danica.

**Comment by moontrip** [comment]
> Looks good. Nice colormap solution to the tricky problem of all the series being interleaved.
> 
> Might be nice to put `', Smoothed mean'`, `', 95% Confidence interval'` etc into "global" constants for this file, so that future edits will only need to be done in one place.
> 
> There is potentially a problem with some real-life variables having actual "No data" values from the database/back end. I don't know one off the top of my head, but will ask Sheena/Danica.

@bobular thank you for your review and suggestions! 👍 As for global constants, do you mean that the computation of the legendLabel is separated as a function/util ? It would be great if you can elaborate it or show an example for me.

I could see the potential issue if actual data having "No data" text in the variable name, but the only problematic case is it has exact "No data" as the string.includes() is case-sensitive - not quite sure if there still exists the exception, though

**Comment by bobular** [comment]
Hi DK - I pushed a minor commit with the string literals as constants - no biggie.
One small step towards internationalisation (which we don't have to do, haha).

I have found an example of 'No data' as actual data and I'm going to spend half an hour to see if I can mitigate the problem now.  Enjoy your weekend!

**Review by bobular — APPROVED** [review]
I will make another "'No data' from database" ticket because it requires discussion with the data curation team too.  All looks good here. Thanks.
Here's the new ticket: https://github.com/VEuPathDB/EdaNewIssues/issues/83

**Comment by moontrip** [comment]
@bobular Aha I got your point now with your changes - I was not quite sure of the exact meaning you had in mind. Thank you for changing it 👍 

I also see your suggestions in the other ticket as well but if I understand correctly your #2 solution may not fix it all cases as "No data" data can happen at OverlayVariable as well. I will comment it too, just in case

**Pull request merged**